lesson20/main.js

```diff
@@ -1 +1,146 @@
+const jsonUrl = "https://myjson.dit.upm.es/api/bins/cpch";
+const tableWrap = document.getElementById("js-table-wrapper");
 
+const createElementWithClassName = (element, name) => {
+  const createdElement = document.createElement(element);
+  createdElement.classList.add(name);
+  return createdElement;
+};
+
+const createLoader = () => {
+  const loader = document.createElement("div");
+  const loaderImage = document.createElement("img");
+  loader.id = "js-loader";
+  loaderImage.src = "img/loading-circle.gif";
+  loader.appendChild(loaderImage);
+  return loader;
+};
+
+const loading = () => {
+  const loader = document.getElementById("js-loader");
+  loader.classList.add("loading");
+};
+
+const loaded = () => {
+  const loader = document.getElementById("js-loader");
+  loader.classList.add("loaded");
+};
+
+const fetchDataInSecond = (sec, jsonUrl) => {
+  return new Promise((resolve, reject) => {
+    setTimeout(() => {
+      resolve(fetch(jsonUrl));
+    }, sec);
+  });
+};
+
+const fetchUsersData = async () => {
+  try {
+    const response = await fetchDataInSecond(3000, jsonUrl);
+    const json = await response.json();
+    return json.data;
+  } catch (e) {
+    tableWrap.textContent = "データの取得ができませんでした。";
+    console.error(e);
+  } finally {
+    loaded();
+  }
+};
+
+/**
+ * Rendering table using users data from JSON file
+ * @param {Object} usersData array of users data
+ */
+const renderTable = (usersData) => {
+  const table = createTable();
+  table.appendChild(createTableHeader(usersData));
+  table.appendChild(createTableData(usersData));
+  tableWrap.appendChild(table);
+};
+
+const createTable = () => {
+  const table = createElementWithClassName("table", "users-table");
+  table.id = "js-table";
+  return table;
+};
+
+/**
+ *Creating table header with key of usersData
+ * @param {Array} usersData The Array of usersData
+ */
+ const createTableHeader = (usersData) => {
+  const trOfThead = document.createElement("tr");
+  const fragment = document.createDocumentFragment();
+  /**
+   *Extract key from usersData
+   * @param {Object} usersData[0] value from one of userData in the object
+   * @param {String} key key of userData
+   */
+  Object.keys(usersData[0]).forEach((key) => {
+    const th = createElementWithClassName("th", "users-table__th");
+    th.textContent = formingTableHeaderNameWithKey(key);
+    fragment.appendChild(th);
+  });
+  trOfThead.appendChild(fragment);
+  return trOfThead;
+};
```

Comment:
IMO
I think I could cut it out here.
```suggestion
const createTableHeader = (usersData) => {
  const trOfThead = document.createElement("tr");
  const fragment = document.createDocumentFragment();
  AddTextContentToTheads(usersData, fragment);
  trOfThead.appendChild(fragment);
  return trOfThead;
};

/**
 *Extract key from usersData
 * @param {Object} usersData[0] value from one of userData in the object
 * @param {String} key key of userData
 */
const AddTextContentToTheads = (usersData, fragment) => {
  Object.keys(usersData[0]).forEach((key) => {
    const th = createElementWithClassName("th", "users-table__th");
    th.textContent = formingTableHeaderNameWithKey(key);
    fragment.appendChild(th);
  });
};
```

Reply:
It became more readable! Thank you :)
[2f49e51](https://github.com/yuka830/js-lessons/pull/26/commits/2f49e510284c6b7fe81c680e56de06e94a240df1)